### PR TITLE
Optimize GEMM Prefetch Block

### PR DESCRIPTION
**Pull request by LiyangLingIntel**
The pull request include three main changes:
1. move inner loop prefetch to the position before dpas after 2d load.
2. similar to https://github.com/intel/intel-xpu-backend-for-triton/pull/1225, change default prefetch distance to 2 which is better for XPU gemm cases. 
3. use env flag `TRITON_INTEL_SPLIT_BARRIER` to control the default behavior of `inject_split_barriers`. In practice, it show better performance for some cases when `TRITON_INTEL_SPLIT_BARRIER=1`. For this pull request we still keep it False by default.

Move `prefetch_distance` and `split_barriers` to `XPUOptions`, so we can maintain them like `num_stages` when initialize XPUBackend.

### CODE REVIEW

File: third_party/intel/backend/compiler.py [hunk 1]
```diff
@@ -79,8 +81,7 @@ def make_ttgir(mod, metadata, opt, device_arch):
             # FIXME: Use a better way to check if prefetch instructions are supported once available.
             # Prefetch instruction is not available in older drivers.
             if Version(metadata["target"].arch['driver_version']) > Version("1.3.28202"):
-                inject_split_barriers = False
-                intel.passes.ttgpuir.add_prefetch_block(pm, opt.num_stages, inject_split_barriers)
+                intel.passes.ttgpuir.add_prefetch_block(pm, opt.prefetch_distance, opt.inject_split_barrier)
```
Comment (whitneywhtsang) [inline third_party/intel/backend/compiler.py:84]: Why not use `opt.num_stages` directly?

File: third_party/intel/lib/TritonIntelGPUTransforms/PrefetchBlock.cpp [hunk 2]
```diff
@@ -138,18 +138,20 @@ class PrefetchBlockPass
 public:
   /// Groups information for a candidate load.
   struct LoadInfo {
-    LoadInfo(tt::AdvanceOp advance, SmallVector<Value> offsets,
+    LoadInfo(tt::DotOp dot, tt::AdvanceOp advance, SmallVector<Value> offsets,
```
Comment (Dewei-Wang-sh) [inline third_party/intel/lib/TritonIntelGPUTransforms/PrefetchBlock.cpp:141]: this pass is meant to be general, not bound to dot op. 
so try to move prefetch after load.

File: third_party/intel/backend/compiler.py (hunk 1 above)
Comment (LiyangLingIntel) [reply third_party/intel/backend/compiler.py:84]: In my understanding `opt.num_stages` is used by `SoftwarePipeline` pass, it's designed for scheduling inner loop operations by stages, on semantic it's different with `prefetch_distance`. Use `opt.prefetch_distance` here is to reduce confusion.

File: third_party/intel/lib/TritonIntelGPUTransforms/PrefetchBlock.cpp (hunk 2 above)
Comment (LiyangLingIntel) [reply third_party/intel/lib/TritonIntelGPUTransforms/PrefetchBlock.cpp:141]: OK, I'll update it soon. Thanks for the suggestion.

File: third_party/intel/backend/compiler.py [hunk 3]
```diff
@@ -48,7 +48,9 @@ class XPUOptions:
     extern_libs: dict = None
     debug: bool = False
     backend_name: str = 'intel'
-    isBlockPtrEnabled: bool = os.environ.get("TRITON_INTEL_ENABLE_BLOCK_PTR", "0") == "1"
+    is_block_ptr_enabled: bool = os.environ.get("TRITON_INTEL_ENABLE_BLOCK_PTR", "0") == "1"
+    prefetch_distance: int = 2
+    inject_split_barrier: bool = os.getenv("TRITON_INTEL_SPLIT_BARRIER", "0") == "1"
```
Comment (whitneywhtsang) [inline third_party/intel/backend/compiler.py:53]: Let's try to be consistent with our usages of `os.getenv` and `os.environ.get`. Unless there is a reason to be different?

Comment (LiyangLingIntel) [reply third_party/intel/backend/compiler.py]: `os.getenv` is a shortcut to `os.environ.get` with `None` default value. Change them to `os.environ.get` for consistency.
Thanks for the suggestion.

**Review by Dewei-Wang-sh — APPROVED** [review]
lgtm

**Review by whitneywhtsang — APPROVED** [review]

File: third_party/intel/backend/compiler.py [hunk 4]
```diff
@@ -48,14 +48,16 @@ class XPUOptions:
     extern_libs: dict = None
     debug: bool = False
     backend_name: str = 'intel'
-    isBlockPtrEnabled: bool = os.environ.get("TRITON_INTEL_ENABLE_BLOCK_PTR", "0") == "1"
+    is_block_ptr_enabled: bool = os.environ.get("TRITON_INTEL_ENABLE_BLOCK_PTR", "0") == "1"
+    prefetch_distance: int = 2
+    inject_split_barrier: bool = os.environ.get("TRITON_INTEL_SPLIT_BARRIER", "0") == "1"
```
Comment (etiotto) [inline third_party/intel/backend/compiler.py:53]: I do not think this is necessary. Introducing too many tuning flags is confusing. What is the difference in performance for cases that benefit from injecting the split barrier ? And how much does injecting a split barrier by default hurts for some configurations?

File: third_party/intel/backend/compiler.py (hunk 1 above)
Comment (etiotto) [reply third_party/intel/backend/compiler.py:84]: We do not need both num_stages and prefetch_distance. Remove "num_stages" then.

File: third_party/intel/backend/compiler.py [hunk 5]
```diff
@@ -122,6 +123,8 @@ def parse_target(self, tgt_prop) -> dict:
     def parse_options(self, opts) -> Any:
         args = {k: opts[k] for k in XPUOptions.__dataclass_fields__.keys() if k in opts}
         args["allow_fp8e4nv"] = True
+        # Use num_stages to represent prefetch distance to make it tunable.
+        args["prefetch_distance"] = opts.get("num_stages", 2)
```
Comment (etiotto) [inline third_party/intel/backend/compiler.py:127]: remove num_stages please and replace it with prefetch_distance. Having both is not necessary.

**Review by etiotto — CHANGES_REQUESTED** [review]

**Review by whitneywhtsang — COMMENTED** [review]
https://github.com/intel/intel-xpu-backend-for-triton/pull/1256 should be able to address Ettore's comments. @LiyangLingIntel Please check if that PR is ok with you.

File: third_party/intel/backend/compiler.py [hunk 6]
```diff
@@ -48,14 +48,14 @@ class XPUOptions:
...
-    isBlockPtrEnabled: bool = os.environ.get("TRITON_INTEL_ENABLE_BLOCK_PTR", "0") == "1"
+    is_block_ptr_enabled: bool = os.environ.get("TRITON_INTEL_ENABLE_BLOCK_PTR", "0") == "1"
 
     def __post_init__(self):
         default_libdir = Path(__file__).parent / 'lib'
         extern_libs = {} if self.extern_libs is None else dict(self.extern_libs)
         if not extern_libs.get('libdevice', None):
-            extern_libs['libdevice'] = os.getenv("TRITON_LIBDEVICE_PATH",
-                                                 str(default_libdir / 'libsycl-spir64-unknown-unknown.bc'))
+            extern_libs['libdevice'] = os.environ.get("TRITON_LIBDEVICE_PATH",
```
Comment (whitneywhtsang) [inline third_party/intel/backend/compiler.py:57]: Let's change back to use `os.getenv` in this case, to be consistent with nvidia and amd.

**Review by whitneywhtsang — APPROVED** [review]

File: third_party/intel/backend/compiler.py (hunk 6 above)
Comment (LiyangLingIntel) [reply third_party/intel/backend/compiler.py]: Updated.

File: third_party/intel/backend/compiler.py (hunk 4 above)
Comment (LiyangLingIntel) [reply third_party/intel/backend/compiler.py]: There only 2 cases benefit from split barrier with around 20% performance gain, for most cases there are 5-10% performance drop.
I removed `prefetch_distance` and `inject_split_barrier` from this pull request for sake of simplicity. keep use `num_stages` to represent `prefetch_distance` to make the interface consistent with upstream.

File: third_party/intel/backend/compiler.py (hunk 5 above)
Comment (LiyangLingIntel) [reply third_party/intel/backend/compiler.py]: `num_stages` is necessary for `SoftwarePipeline` pass, so I remove the option `prefetch_distance` and keep it as it was. In this way we can make our interface consistent with upstream as well.

File: third_party/intel/backend/compiler.py (hunk 4 above)
Comment (etiotto) [reply third_party/intel/backend/compiler.py]: why 5-10% degradation ? Have you reported the issue to IGC ?

File: third_party/intel/backend/compiler.py (hunk 5 above)
Comment (etiotto) [reply third_party/intel/backend/compiler.py]: ok

**Review by etiotto — APPROVED** [review]

**Pull request merged**